doc/nona.1
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "NONA ""1"""
.TH NONA "1" "2010-12-26" """Version: 2010.5.0""" "HUGIN"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
nona \- Stitch a panorama image
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBnona\fR [options] \fB\-o\fR \fIoutput\fR \fIproject_file\fR (image files)
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBnona\fR uses the transform function from PanoTools, the stitching
itself is quite simple, no seam feathering is done.
.PP
Only the non-antialiasing interpolators of PanoTools are supported.
.PP
The following output formats (\fBn\fR option of PanoTools \fBp\fR script
line) are supported:
.IP "\s-1JPEG\s0, \s-1TIFF\s0, \s-1PNG\s0  : Single image formats without feathered blending" 4
.IX Item "JPEG, TIFF, PNG  : Single image formats without feathered blending"
.PD 0
.IP "TIFF_m          : multiple tiff files" 4
.IX Item "TIFF_m          : multiple tiff files"
.IP "TIFF_multilayer : Multilayer tiff files, readable by The Gimp 2.0" 4
.IX Item "TIFF_multilayer : Multilayer tiff files, readable by The Gimp 2.0"
.PD
.SH "OPTIONS"
.IX Header "OPTIONS"
General options:
.IP "\fB\-c\fR" 4
.IX Item "-c"
Create coordinate images (only TIFF_m output)
.IP "\fB\-v\fR" 4
.IX Item "-v"
Quiet, do not output progress indicators
.IP "\fB\-t\fR \fInum\fR" 4
.IX Item "-t num"
Number of threads to be used (default: number of available cores)
.PP
The following options can be used to override settings in the project file:
.IP "\fB\-i\fR \fInum\fR" 4
.IX Item "-i num"
Remap only image with number num (can be specified multiple times)
.IP "\fB\-m\fR \fIstr\fR" 4
.IX Item "-m str"
Set output file format (\s-1TIFF\s0, TIFF_m, TIFF_multilayer, \s-1EXR\s0, EXR_m)
.IP "\fB\-r\fR \fIldr\fR/\fIhdr\fR" 4
.IX Item "-r ldr/hdr"
Set output mode:
.RS 4
.IP "ldr \- keep original bit depth and response" 4
.IX Item "ldr - keep original bit depth and response"
.PD 0
.IP "hdr \- merge to hdr" 4
.IX Item "hdr - merge to hdr"
.RE
.RS 4
.RE
.IP "\fB\-e\fR \fIexposure\fR" 4
.IX Item "-e exposure"
.PD
Set exposure for ldr mode
.IP "\fB\-p\fR \fI\s-1TYPE\s0\fR" 4
.IX Item "-p TYPE"
Pixel type of the output. Can be one of:
.RS 4
.IP "\s-1UINT8\s0   8 bit unsigned integer" 4
.IX Item "UINT8   8 bit unsigned integer"
.PD 0
.IP "\s-1UINT16\s0  16 bit unsigned integer" 4
.IX Item "UINT16  16 bit unsigned integer"
.IP "\s-1INT16\s0   16 bit signed integer" 4
.IX Item "INT16   16 bit signed integer"
.IP "\s-1UINT32\s0  32 bit unsigned integer" 4
.IX Item "UINT32  32 bit unsigned integer"
.IP "\s-1INT32\s0   32 bit signed integer" 4
.IX Item "INT32   32 bit signed integer"
.IP "\s-1FLOAT\s0   32 bit floating point" 4
.IX Item "FLOAT   32 bit floating point"
.RE
.RS 4
.RE
.IP "\fB\-z\fR" 4
.IX Item "-z"
.PD
Set compression type. Possible options for tiff output:
.RS 4
.IP "\s-1NONE\s0      no compression" 4
.IX Item "NONE      no compression"
.PD 0
.IP "\s-1LZW\s0       \s-1LZW\s0 compression" 4
.IX Item "LZW       LZW compression"
.IP "\s-1DEFLATE\s0   deflate compression" 4
.IX Item "DEFLATE   deflate compression"
.RE
.RS 4
.RE
.PD
.SH "AUTHORS"
.IX Header "AUTHORS"
Written by Pablo d'Angelo. Also contains contributions from Douglas Wilkins, Ippei Ukai, Ed Halley, Bruno Postle, Gerry Patterson and Brent Townshend.
.PP
This man page was written by Cyril Brulebois
<cyril.brulebois@enst\-bretagne.fr> and is licensed under the same
terms as the hugin package itself.
.SH "POD ERRORS"
.IX Header "POD ERRORS"
Hey! \fBThe above document had some coding errors, which are explained below:\fR
.IP "Around line 129:" 4
.IX Item "Around line 129:"
You forgot a '=back' before '=head1'


NEWS
Newsworthy changes in 0.7 beta 5
--------------------------------
* much improved vignetting and exposure correction,
  http://hugin.sf.net/tech
* ability to create HDR images during stitching
  (still extremly memory hungry, will be improved
   in the next version).
* fixed bad HFOV/VFOV limits of some projections
* OSX build should not crash all the time
* Added "Apply template" function, to apply the
  settings of a previously saved project to
  the current images.
* enabled "load lens data" button in dialog asking
  for focal length/HFOV.

Newsworthy changes in 0.7 beta 4
-------------------------
* update panotools to latest version
* Albers equal area conic and Miller cylindrical projection
* Support for straight line control points
* Fixed running of Autopano-SIFT integration on windows
* (linux) work around bugs in wxWidgets (fixes double deletion
  of control points when pressing the Del key)
* Force even values for width for 360 degree equirectangular
  panoramas.

Newsworthy changes in 0.7 beta 3
-------------------------
* Assistant panel, enables 3 click stitching (well, almost ;-)
* Center and rotate panorama by mouse clicks in preview window.
  Also added a numerical transform dialog
* automatic straighten
* support for Lambert and Lambert Azimuthal projections
* separate crop tab replaces buggy crop dialog
* crop is saved in lens .ini files
* Rotated display of images in control point selector, depending
  on the roll value
* allow images with and without alpha channels in the same project
* option to open final panorama in image editor after stitching
* present output of command line utils in nice window (unix)
* many small bugfixes
* removed image center dialog


Newsworthy changes in 0.6.1
---------------------------
* Autopano-SIFT integration was broken in 0.6 on Unix and OSX.
* Support for displaying HDR files within hugin, using log or gamma mapping
* .hdr file support (tif HDR files have been always useable)
* reduced memory usage, better image caching strategy in GUI.
* fixed crashes and hangs when using the preview auto update and avoid
  unnessecary remapping of images.
* vignetting correction estimation speed up.
* nona and nona_gui can be used to do template stitching, by specifying
  images on the command line (or dnd shortcut under windows)
* distribute fulla.exe with windows package (it was accidentally
  not included in the 0.6 package)
* universal build for Intel Macs, fixed hang after enblend or autopano-sift
  call
* projection can be changed in preview.
* removed all references to malicious panotools.info website
* various smaller bugfixes

Newsworthy changes in 0.6
-------------------------
* updated vigra to version 1.4. All sourcecode of hugin is now
  released under approved open source licences. This means there
  is no problem with including hugin in debian and ubuntu anymore.
* Copy ICC profiles from input images to output
* Added vignetting correction and brighness correction
  during stitching.
  Vignetting can be corrected by addition or division.
  The correction can be done with a flatfield image or an
  radial polynomial
* Functionality to estimate the vignetting polynomial
  directly from the overlapping images. This is an experimental feature.
* Multithreaded remapping. hugin and nona fully utilize multiple
  CPUs or multi core CPUs during preview generation and stitching
* support for dcraw 16 bit .ppm files
* Disabled PTOptimizer progress display on OSX. This should fix
  the crashes some people have experienced.
* New command line tool (fulla), for distortion, TCA and vignetting correction.
  Can use PTLens database (old format only).
* reorganized GUI layout.
* Czech translation added

Newsworthy changes in 0.5
-------------------------
* Added Catalan, Chinese and Hungarian translation
* Support TIFF compressions: NONE, DEFLATE, LZW
* option to create cropped tiff files for usage with enblend 2.4
* Mac OSX: fixed bug during calling PTOptimizer executable
* Windows only:
    hugin is shipped with the pano12.dll build with MSVC.
    This might lead to crashes when using PTStitcher.exe
    In that case replace pano12.dll with pano12_for_usage_with ptstitcher.dll

Newsworthy changes in 0.5 rc2
-----------------------------
* Removed all non-opensource sourcecode from CVS
* Fixed various crashes
* Japanese, Russian, Brazilian and Dutch translation
* Proper default arguments for Autopano-SIFT on Windows
  (users of hugin 0.5 r1 or less need to restore the default
   config)

Newsworthy changes in 0.5 rc1
-----------------------------
* Options on the menu and toolbar that are irrelevant
  are now disabled.
* enblend can now be called from nona even if it has 
  spaces in the path.

Newsworthy changes in 0.5 beta7
-------------------------------
* Updated translations
* Fixes to the build system on linux and windows

Newsworthy changes in 0.5 beta6
-------------------------------
* All files are now in source tarball
* French translation now complete

Newsworthy changes in 0.5 beta5
-------------------------------

* Support for modifying the size of the right/left panes in many tabs.
* All controls can now be reached on smaller displays as many areas 
  of the interface are now scrollable.
* All image control buttons in the preview tab can now be accessed, 
  allowing support for far more images.

* OSX build now supports unicode.

* Enblend arguments specified in preferences are now passed on correctly.
* Multiple tiff file outputs are no longer deleted incorrectly.
* Simultaneous scrolling in control points editor tabs respects borders properly.


src/tools/nona.cpp
// -*- c-basic-offset: 4 -*-
/** @file nona.cpp
 *
 *  @brief a simple test stitcher
 *
 *  @author Pablo d'Angelo <pablo.dangelo@web.de>
 *
 *  $Id$
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public
 *  License as published by the Free Software Foundation; either
 *  version 2 of the License, or (at your option) any later version.
 *
 *  This software is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public
 *  License along with this software; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

#include <hugin_config.h>
#include <hugin_version.h>
#include <fstream>
#include <sstream>

#include <algorithm>
#include <cctype>
#include <string>

#include <vigra/error.hxx>
#include <vigra/impex.hxx>

#ifdef WIN32
 #include <getopt.h>
#else
 #include <unistd.h>
#endif

#include <hugin_basic.h>
#include <hugin_utils/platform.h>
#include <algorithms/nona/NonaFileStitcher.h>
#include <vigra_ext/MultiThreadOperations.h>

#include <tiffio.h>

#if !defined Hugin_shared || !defined _WINDOWS
#define GLEW_STATIC
#endif
#include <GL/glew.h>
#ifdef __APPLE__
  #include <GLUT/glut.h>
#else
  #include <GL/glut.h>
#endif

using namespace vigra;
using namespace HuginBase;
using namespace hugin_utils;
using namespace std;

GLuint GlutWindowHandle;

static void usage(const char * name)
{
    cerr << name << ": stitch a panorama image" << std::endl
    << std::endl
    << "nona version " << DISPLAY_VERSION << std::endl
    << std::endl
    << " It uses the transform function from PanoTools, the stitching itself" << std::endl
    << " is quite simple, no seam feathering is done." << std::endl
    << " only the non-antialiasing interpolators of panotools are supported" << std::endl
    << std::endl
    << " The following output formats (n option of panotools p script line)" << std::endl
    << " are supported:"<< std::endl
    << std::endl
    << "  JPG, TIFF, PNG  : Single image formats without feathered blending:"<< std::endl
    << "  TIFF_m          : multiple tiff files"<< std::endl
    << "  TIFF_multilayer : Multilayer tiff files, readable by The Gimp 2.0" << std::endl
    << std::endl
    << "Usage: " << name  << " [options] -o output project_file (image files)" << std::endl
    << "  Options: " << std::endl
    << "      -c         create coordinate images (only TIFF_m output)" << std::endl
    << "      -v         quiet, do not output progress indicators" << std::endl
    << "      -t num     number of threads to be used (default: nr of available cores)" << std::endl
    << "      -g         perform image remapping on the GPU" << std::endl
    << std::endl
    << "  The following options can be used to override settings in the project file:" << std::endl
    << "      -i num     remap only image with number num" << std::endl
    << "                   (can be specified multiple times)" << std::endl
    << "      -m str     set output file format (TIFF, TIFF_m, TIFF_multilayer, EXR, EXR_m)" << std::endl
    << "      -r ldr/hdr set output mode." << std::endl
    << "                   ldr  keep original bit depth and response" << std::endl
    << "                   hdr  merge to hdr" << std::endl
    << "      -e exposure set exposure for ldr mode" << std::endl
    << "      -p TYPE    pixel type of the output. Can be one of:" << std::endl
    << "                  UINT8   8 bit unsigned integer" << std::endl
    << "                  UINT16  16 bit unsigned integer" << std::endl
    << "                  INT16   16 bit signed integer" << std::endl
    << "                  UINT32  32 bit unsigned integer" << std::endl
    << "                  INT32   32 bit signed integer" << std::endl
    << "                  FLOAT   32 bit floating point" << std::endl
    << "      -z         set compression type." << std::endl
    << "                  Possible options for tiff output:" << std::endl
    << "                   NONE      no compression" << std::endl
    << "                   PACKBITS  packbits compression" << std::endl
    << "                   LZW       lzw compression" << std::endl
    << "                   DEFLATE   deflate compression" << std::endl
    << std::endl;
}

/** Try to initalise GLUT and GLEW, and create an OpenGL context for GPU stitching.
 * OpenGL extensions required by the GPU stitcher (-g option) are checked here.
 * @return true if everything went OK, false if we can't use GPGPU mode.
 */
static bool initGPU(int *argcp,char **argv) {
    glutInit(argcp,argv);
    glutInitDisplayMode(GLUT_SINGLE | GLUT_RGBA | GLUT_ALPHA);
    GlutWindowHandle = glutCreateWindow("nona");

    int err = glewInit();
    if (err != GLEW_OK) {
        cerr << "nona: an error occured while setting up the GPU:" << endl;
        cerr << glewGetErrorString(err) << endl;
        cerr << "nona: Switching to CPU calculation." << endl;
        glutDestroyWindow(GlutWindowHandle);
        return false;
    }

    cout << "nona: using graphics card: " << glGetString(GL_VENDOR) << " " << glGetString(GL_RENDERER) << endl;

    GLboolean has_arb_fragment_shader = glewGetExtension("GL_ARB_fragment_shader");
    GLboolean has_arb_vertex_shader = glewGetExtension("GL_ARB_vertex_shader");
    GLboolean has_arb_shader_objects = glewGetExtension("GL_ARB_shader_objects");
    GLboolean has_arb_shading_language = glewGetExtension("GL_ARB_shading_language_100");
    GLboolean has_arb_texture_rectangle = glewGetExtension("GL_ARB_texture_rectangle");
    GLboolean has_arb_texture_border_clamp = glewGetExtension("GL_ARB_texture_border_clamp");
    GLboolean has_arb_texture_float = glewGetExtension("GL_ARB_texture_float");

    if (!(has_arb_fragment_shader && has_arb_vertex_shader && has_arb_shader_objects && has_arb_shading_language && has_arb_texture_rectangle && has_arb_texture_border_clamp && has_arb_texture_float)) {
        const char * msg[] = {"false", "true"};
        cerr << "nona: extension GL_ARB_fragment_shader = " << msg[has_arb_fragment_shader] << endl;
        cerr << "nona: extension GL_ARB_vertex_shader = " << msg[has_arb_vertex_shader] << endl;
        cerr << "nona: extension GL_ARB_shader_objects = " << msg[has_arb_shader_objects] << endl;
        cerr << "nona: extension GL_ARB_shading_language_100 = " << msg[has_arb_shading_language] << endl;
        cerr << "nona: extension GL_ARB_texture_rectangle = " << msg[has_arb_texture_rectangle] << endl;
        cerr << "nona: extension GL_ARB_texture_border_clamp = " << msg[has_arb_texture_border_clamp] << endl;
        cerr << "nona: extension GL_ARB_texture_float = " << msg[has_arb_texture_float] << endl;
        cerr << "nona: This graphics system lacks the necessary extensions for -g." << endl;
        cerr << "nona: Switching to CPU calculation." << endl;
        glutDestroyWindow(GlutWindowHandle);
        return false;
    }

    return true;
}

static bool wrapupGPU() {
    glutDestroyWindow(GlutWindowHandle);
    return true;
}

int main(int argc, char *argv[])
{
    
    // parse arguments
    const char * optstring = "z:cho:i:t:m:p:r:e:vg";
    int c;
    
    opterr = 0;
    
    int nThread = getCPUCount();
    if (nThread < 0) nThread = 1;
    bool doCoord = false;
    UIntSet outputImages;
    string basename;
    string outputFormat;
    bool overrideOutputMode = false;
    std::string compression;
    PanoramaOptions::OutputMode outputMode = PanoramaOptions::OUTPUT_LDR;
    bool overrideExposure = false;
    double exposure=0;
    int verbose = 0;
    int useGPU = 0;
    string outputPixelType;
    
    while ((c = getopt (argc, argv, optstring)) != -1)
    {
        switch (c) {
            case 'o':
                basename = optarg;
                break;
            case 'c':
                doCoord = true;
                break;
            case 'i':
                outputImages.insert(atoi(optarg));
                break;
            case 'm':
                outputFormat = optarg;
                break;
            case 'p':
                outputPixelType = optarg;
                break;
            case 'r':
                if (string(optarg) == "ldr") {
                    overrideOutputMode = true;
                    outputMode = PanoramaOptions::OUTPUT_LDR;
                } else if (string(optarg) == "hdr") {
                    overrideOutputMode = true;
                    outputMode = PanoramaOptions::OUTPUT_HDR;
                } else {
                    usage(argv[0]);
                    return 1;
                }
                break;
            case 'e':
                overrideExposure = true;
                exposure = atof(optarg);
                break;
            case '?':
            case 'h':
                usage(argv[0]);
                return 0;
            case 't':
                nThread = atoi(optarg);
                break;
            case 'v':
                ++verbose;
                break;
            case 'z':
                compression = optarg;
                std::transform(compression.begin(), compression.end(), compression.begin(), (int(*)(int)) toupper);
                break;
            case 'g':
                useGPU = 1;
                break;
            default:
		usage(argv[0]);
                abort ();
        }
    }

    if (basename == "" || argc - optind <1) {
        usage(argv[0]);
        return 1;
    }
    unsigned nCmdLineImgs = argc -optind -1;

    if (nThread == 0) nThread = 1;
    vigra_ext::ThreadManager::get().setNThreads(nThread);

    const char * scriptFile = argv[optind];

    // suppress tiff warnings
    TIFFSetWarningHandler(0);

    AppBase::ProgressDisplay* pdisp = NULL;
    if(verbose > 0)
        pdisp = new AppBase::StreamProgressDisplay(cout);
    else
        pdisp = new AppBase::DummyProgressDisplay;

    Panorama pano;
    ifstream prjfile(scriptFile);
    if (prjfile.bad()) {
        cerr << "could not open script : " << scriptFile << std::endl;
        exit(1);
    }
    pano.setFilePrefix(hugin_utils::getPathPrefix(scriptFile));
    AppBase::DocumentData::ReadWriteError err = pano.readData(prjfile);
    if (err != AppBase::DocumentData::SUCCESSFUL) {
        cerr << "error while parsing panos tool script: " << scriptFile << std::endl;
        exit(1);
    }

    if ( nCmdLineImgs > 0) {
        if (nCmdLineImgs != pano.getNrOfImages()) {
            cerr << "Incorrect number of images specified on command line\nProject required " << pano.getNrOfImages() << " but " << nCmdLineImgs << " where given" << std::endl;
            exit(1);
        }
        for (unsigned i=0; i < pano.getNrOfImages(); i++) {
            pano.setImageFilename(i, argv[optind+i+1]);
        }

    }
    PanoramaOptions  opts = pano.getOptions();

    if (compression.size() > 0) {
        opts.tiffCompression=compression;
    }

    // save coordinate images, if requested
    opts.saveCoordImgs = doCoord;
    if (outputFormat == "TIFF_m") {
        opts.outputFormat = PanoramaOptions::TIFF_m;
    } else if (outputFormat == "TIFF") {
        opts.outputFormat = PanoramaOptions::TIFF;
    } else if (outputFormat == "TIFF_multilayer") {
        opts.outputFormat = PanoramaOptions::TIFF_multilayer;
    } else if (outputFormat == "EXR_m") {
        opts.outputFormat = PanoramaOptions::EXR_m;
    } else if (outputFormat == "EXR") {
        opts.outputFormat = PanoramaOptions::EXR;
    } else if (outputFormat != "") {
        cerr << "Error: unknown output format: " << outputFormat << endl;
        return 1;
    }

    if (outputPixelType.size() > 0) {
        opts.outputPixelType = outputPixelType;
    }
    
    if (overrideOutputMode) {
        opts.outputMode = outputMode;
    }
    
    if (overrideExposure) {
        opts.outputExposureValue = exposure;
    }

    if(outputImages.size()==0) 
    {
        outputImages = pano.getActiveImages();
    }
    else
    {
        UIntSet activeImages=pano.getActiveImages();
        for(UIntSet::const_iterator it=outputImages.begin(); it!=outputImages.end();it++)
        {
            if(!set_contains(activeImages,*it))
            {
                std::cerr << "The project file does not contains an image with number " << *it << std::endl;
                return 1;
            };
        };
    };
    if(outputImages.size()==0)
    {
        std::cout << "Project does not contain active images." << std::endl
            << "Nothing to do for nona." << std::endl;
        return 0;
    };
    if(useGPU)
    {
        switch(opts.getProjection())
        {
            // the following projections are not supported by nona-gpu
            case HuginBase::PanoramaOptions::ARCHITECTURAL:
            case HuginBase::PanoramaOptions::BIPLANE:
            case HuginBase::PanoramaOptions::TRIPLANE:
            case HuginBase::PanoramaOptions::EQUISOLID:
            case HuginBase::PanoramaOptions::ORTHOGRAPHIC:
            case HuginBase::PanoramaOptions::PANINI:
            case HuginBase::PanoramaOptions::EQUI_PANINI:
            case HuginBase::PanoramaOptions::GENERAL_PANINI:
                useGPU=false;
                std::cout << "Nona-GPU does not support this projection. Switch to CPU calculation."<<std::endl;
                break;
        };
        for(UIntSet::const_iterator it=outputImages.begin(); it!=outputImages.end();it++)
        {
            const SrcPanoImage & img = pano.getImage(*it);
            if(img.getX()!=0 || img.getY()!=0 || img.getZ()!=0)
            {
                useGPU=false;
                std::cout << "Nona-GPU does not support the translation parameters. Switch to CPU calculation." << std::endl;
                break;
            };
        };
    };
    
    DEBUG_DEBUG("output basename: " << basename);
    
    try {
        if (useGPU) {
            useGPU = initGPU(&argc, argv);
        }
        opts.remapUsingGPU = useGPU;
        pano.setOptions(opts);

        // stitch panorama
        NonaFileOutputStitcher(pano, pdisp, opts, outputImages, basename).run();
        // add a final newline, after the last progress message
        if (verbose > 0) {
            cout << std::endl;
        }

        if (useGPU) {
            wrapupGPU();
        }

    } catch (std::exception & e) {
        cerr << "caught exception: " << e.what() << std::endl;
        return 1;
    }
    
    if(pdisp != NULL)
        delete pdisp;

    return 0;
}
